File: api/src/handlers/starter_templates/chess_skeleton.go
// Chess starter skeleton in Go. Wire protocol: judge/protocols/wire.md.
// Game spec: judge/protocols/chess.md.
//
// Replace chooseMove with your engine. Returning "" skips the turn.
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type state struct {
	moveCount int
	// TODO: board, side-to-move, my_seat, last_move, ...
}

func chooseMove(s *state) string {
	// TODO: return e.g. "e2 e4 -".
	return ""
}

func main() {
	r := bufio.NewScanner(os.Stdin)
	r.Buffer(make([]byte, 0, 1024), 64*1024)
	w := bufio.NewWriter(os.Stdout)
	defer w.Flush()
	var s state

	for r.Scan() {
		tokens := strings.Fields(r.Text())
		if len(tokens) < 3 || tokens[0] != "EVENT" {
			continue
		}
		switch tokens[2] {
		case "GAME_STARTED":
			if mv := chooseMove(&s); mv != "" {
				fmt.Fprintf(w, "ACT MOVE %s\n", mv)
				w.Flush()
			}
		case "MOVE":
			s.moveCount++
			if mv := chooseMove(&s); mv != "" {
				fmt.Fprintf(w, "ACT MOVE %s\n", mv)
				w.Flush()
			}
		case "WINNER", "DRAW":
			return
		}
	}
}

File: api/src/handlers/starter_templates/chess_skeleton.rs
// Chess starter skeleton. Wire protocol: judge/protocols/wire.md.
// Game spec: judge/protocols/chess.md.
//
// You play `MOVE <from> <to> [promo]` where squares are algebraic
// (`a1`..`h8`). Player 0 = white (moves first), player 1 = black.
// Bots are not told their seat; derive it from the order of MOVE
// events you see before GAME_STARTED finishes.
//
// Replace the body of `choose_move` with your engine. Returning
// `None` skips a turn — useful while you scaffold.

use std::io::{self, BufRead, Write};

#[derive(Default)]
struct State {
    // TODO: track board, side-to-move, my_seat, last_move, etc.
    move_count: u32,
}

fn choose_move(_state: &State) -> Option<String> {
    // TODO: return e.g. Some("e2 e4 -".into()).
    None
}

fn main() {
    let stdin = io::stdin();
    let mut stdout = io::stdout();
    let mut state = State::default();

    for line in stdin.lock().lines() {
        let line = match line {
            Ok(l) => l,
            Err(_) => break,
        };
        let mut tok = line.split_whitespace();
        if tok.next() != Some("EVENT") {
            continue;
        }
        let _seq = tok.next();
        let kind = match tok.next() {
            Some(k) => k,
            None => continue,
        };
        let payload: Vec<&str> = tok.collect();

        match kind {
            "GAME_STARTED" => {
                if let Some(act) = choose_move(&state) {
                    let _ = writeln!(stdout, "ACT MOVE {act}");
                    let _ = stdout.flush();
                }
            }
            "MOVE" => {
                state.move_count += 1;
                // payload = [pid, from, to, flag]
                let _ = payload;
                if let Some(act) = choose_move(&state) {
                    let _ = writeln!(stdout, "ACT MOVE {act}");
                    let _ = stdout.flush();
                }
            }
            "WINNER" | "DRAW" => break,
            _ => {}
        }
    }
}
